language/move-prover/bytecode/src/borrow_analysis_v2.rs

```diff
@@ -0,0 +1,622 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! Data flow analysis computing borrow information for preparation of memory_instrumentation.
+
+use crate::{
+    dataflow_analysis::{AbstractDomain, DataflowAnalysis, JoinResult, TransferFunctions},
+    function_target::{FunctionData, FunctionTarget},
+    function_target_pipeline::{FunctionTargetProcessor, FunctionTargetsHolder},
+    livevar_analysis::LiveVarAnnotation,
+    stackless_bytecode::{AssignKind, BorrowNode, Bytecode, Operation},
+    stackless_control_flow_graph::StacklessControlFlowGraph,
+};
+use itertools::Itertools;
+use move_model::{
+    ast::TempIndex,
+    model::{FunctionEnv, QualifiedId},
+};
+use std::collections::{BTreeMap, BTreeSet};
+use vm::file_format::CodeOffset;
+
+#[derive(Debug, Clone, Eq, Ord, PartialEq, PartialOrd)]
+pub enum BorrowEdge {
+    Weak,
+    Strong(Option<usize>),
+}
+
+impl std::fmt::Display for BorrowEdge {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            BorrowEdge::Weak => write!(f, "*"),
+            BorrowEdge::Strong(edge) => match edge {
+                Option::None => write!(f, ""),
+                Option::Some(offset) => write!(f, "{}", offset),
+            },
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, PartialOrd, Eq, Default)]
+pub struct BorrowInfo {
+    /// Contains the nodes which are alive. This excludes nodes which are alive because
+    /// other nodes which are alive borrow from them.
+    live_nodes: BTreeSet<BorrowNode>,
+
+    /// Contains the nodes which are unchecked regards their pack/unpack invariant.
+    /// These are nodes derived from &mut parameters of private functions for which we do
+    /// not perform pack/unpack.
+    unchecked_nodes: BTreeSet<BorrowNode>,
+
+    /// Contains the nodes which have been updated via a Splice operation.
+    spliced_nodes: BTreeSet<BorrowNode>,
+
+    /// Contains the nodes which have been moved via a move instruction.
+    moved_nodes: BTreeSet<BorrowNode>,
+
+    /// Forward borrow information.
+    borrowed_by: BTreeMap<BorrowNode, BTreeMap<BorrowNode, BTreeSet<BorrowEdge>>>,
```

Comment:
I was expecting the type of borrowed_by to be 

BTreeMap<BorrowNode, BTreeMap<BorrowNode, BorrowEdge>>.

Reply:
I see. I just wasn't 100% sure it wasn't possible that a node couldn't borrow two fields which is why I have it encoded this way.

Reply:
I am not sure either.  We should probably discuss.  We have to think about the join.